add ci artifact registry clean up policy

sailorlqh opened the pull request:
fixes: #151 

The artifact registry clean up doc can be found here: https://cloud.google.com/artifact-registry/docs/repositories/cleanup-policy

Currently I didn't found terraform support for adding clean up policy for artifact registry. 

I have tested the clean up policy with my own test project, and the old images are already deleted.

If we don't want to do it this way, we can wait until terraform supports this, or manually run command like `gcloud artifact registry delete` when cleaning up the ci infra.

sailorlqh commented on .github/workflows/ci.yml:
```diff
@@ -126,6 +126,12 @@ jobs:
+          gcloud artifacts repositories set-cleanup-policies ci-images\
+            --project=${{ env.INTEG_TEST_PROJECT_ID }}\
+            --location=us --policy=internal/testhelper/ci-ar-cleanup-policy.json\
+            --dry-run
```
If we think this is a good idea, I will change this flag to --no-dry-run to apply the policy.

capri-xiyue replied:
Do we need to run it for each ci/cd run? I'm wondering should we have another scheduled workflow to set the policy instead of using ci.yml?

sailorlqh replied:
I thought about it, I think making another workflow seems to be a better way. Because it can other prevent people changing the config file and run CI on that branch, which would lead to changing the cleanup policy before the branch got merged.

sailorlqh replied:
I will update the code.

yolocs replied:
This shouldn't be set repeatedly in CI.

capri-xiyue commented on internal/testhelper/ci-ar-cleanup-policy.json:
```diff
@@ -0,0 +1,9 @@
+      "name": "delete-old-ci-images",
+      "action": {"type": "Delete"},
+      "condition": {
+        "olderThan": "1d"
```
Is 1d a little bit short? Probably 3 or 4 days

sailorlqh replied:
changed to 3days.

sailorlqh commented on internal/testhelper/ci-ar-cleanup-policy.json:
```diff
@@ -0,0 +1,9 @@
+[
```
I put this json file in testhelper because I think it's for ci tests.
If we think putting this file to other directories like `artifact-registry/ci-ar-cleanup-policy.json` or echo this file in `ci workflow`, I think that's also doable.

capri-xiyue commented:
Have you configured the roles https://cloud.google.com/artifact-registry/docs/repositories/cleanup-policy#required_roles

sailorlqh commented:
> Have you configured the roles https://cloud.google.com/artifact-registry/docs/repositories/cleanup-policy#required_roles

Not yet, I though it would be better to change the role after we agreed this is the preferred way.

yolocs commented on .github/workflows/update-ci-ar-cleanup-policy.yml:
```diff
@@ -0,0 +1,84 @@
+
+on:
+  workflow_dispatch:
+  pull_request:
+    branches:
+      - 'main'
```
This should only run once. Not on pull request

sailorlqh replied:
I will remove this before merging to main, this is just here so I can test this workflow before merging to main.

sailorlqh commented on .github/workflows/update-ci-ar-cleanup-policy.yml:
```diff
@@ -0,0 +1,84 @@
+          gcloud artifacts repositories set-cleanup-policies ${{ env.DOCKER_REPO }}\
+            --project=${{ env.CI_PROJECT_ID }}\
+            --location=us --policy=${{ env.AR_CONFIG_FILE }}\
+            --dry-run
```
This will be changed to --no-dry-run after testing and before merging to main

sailorlqh commented:
Now the dry run has passed, see: https://github.com/abcxyz/pmap/actions/runs/5885359938, I changed `--dry-run` to `--no-dry-run` and removed trigger event to only `workflow_dispatch`. Once the PR is merged to main, I will manually trigger this workflow to apply the clean up policy.

capri-xiyue commented on .github/workflows/update-ci-ar-cleanup-policy.yml:
```diff
@@ -0,0 +1,81 @@
+          EOT
+          cat ${CONFIG_FILE}
+          echo "AR_CONFIG_FILE=${CONFIG_FILE}" >> $GITHUB_ENV;
+      - name: 'apply_cleanup_config'
```
Follow the github workflow step naming convention.

sailorlqh replied:
Updated.

capri-xiyue commented on .github/workflows/update-ci-ar-cleanup-policy.yml:
```diff
@@ -0,0 +1,81 @@
+        run: |-
+          gcloud artifacts repositories set-cleanup-policies ${{ env.DOCKER_REPO }}\
+            --project=${{ env.CI_PROJECT_ID }}\
+            --location=us --policy=${{ env.AR_CONFIG_FILE }}\
```
nits: for each flag, use one line

sailorlqh replied:
Done.
